engines/router/missionctl/log/resultlog/upi_result_log.go

```diff
@@ -48,36 +53,139 @@ var convertorTableSchema uint32 = converter.TableSchemaV1
 
 var routerNameRegex = regexp.MustCompile(`^[a-zA-Z0-9_\-]+-\d+.[a-zA-Z0-9\-]+$`)
 
-func InitUPIResultLogger(appName string, logger UPILogger) (*UPIResultLogger, error) {
+func InitUPIResultLogger(
+	appName string,
+	loggerType config.ResultLogger,
+	upiLogger UPILogger,
+	resultLogger *ResultLogger) (*UPIResultLogger, error) {
 
-	resultLogger := &UPIResultLogger{
-		logger: logger,
+	upiResultLogger := &UPIResultLogger{
+		upiLogger:          upiLogger,
+		turingResultLogger: resultLogger,
+		loggerType:         loggerType,
 	}
 	if !routerNameRegex.MatchString(appName) {
 		return nil, fmt.Errorf("invalid router name")
 	}
 	s := strings.Split(appName, ".")
 	routerNameWithVersion := s[0]
-	resultLogger.projectName = s[1]
+	upiResultLogger.projectName = s[1]
 
 	i := strings.LastIndex(routerNameWithVersion, "-")
-	resultLogger.routerName = routerNameWithVersion[:i]
+	upiResultLogger.routerName = routerNameWithVersion[:i]
 	// do not include '-'
-	resultLogger.routerVersion = routerNameWithVersion[i+1:]
+	upiResultLogger.routerVersion = routerNameWithVersion[i+1:]
 
-	return resultLogger, nil
-}
+	// check the respective logger has been provided for the config
+	switch loggerType {
+	case config.BigqueryLogger, config.KafkaLogger:
+		if resultLogger == nil {
+			return nil, fmt.Errorf("init upiresultlogger err: required logger not passed for bq/kafka logging")
+		}
+	case config.UPILogger:
+		if upiLogger == nil {
+			return nil, fmt.Errorf("init upiresultlogger err: required logger not passed for upi logging")
+		}
+	case config.NopLogger:
+		if upiLogger == nil && resultLogger == nil {
+			return nil, fmt.Errorf("init upiresultlogger err: required logger not passed for nop logging")
+		}
+	}
 
-func (ul *UPIResultLogger) LogEntry(routerLog *upiv1.RouterLog) error {
-	return ul.logger.write(routerLog)
+	return upiResultLogger, nil
 }
 
 func (ul *UPIResultLogger) LogTuringRouterRequestSummary(
 	header metadata.MD,
 	upiReq *upiv1.PredictValuesRequest,
 	mcRespCh <-chan GrpcRouterResponse,
 ) {
+	if ul.loggerType == config.UPILogger {
+		logRouterLog(header, upiReq, mcRespCh, ul)
+	} else {
+		logTuringResultLog(header, upiReq, mcRespCh, ul)
+	}
+}
+
+func (ul *UPIResultLogger) logEntry(log *upiv1.RouterLog) error {
+	return ul.upiLogger.write(log)
+}
+
+// LogTuringRouterRequestError logs the given turing request id and the error data
+func (ul *UPIResultLogger) LogTuringRouterRequestError(ctx context.Context, err *errors.TuringError) {
+	logger := log.WithContext(ctx)
+	defer func() {
+		_ = logger.Sync()
```

Comment:
What's the purpose of calling logger.Sync() here?

Reply:
my understanding is to flush any buffered log entries, im not sure if we actually need it, but its all over the codebase to do a sync for many logging

Reply:
Per the Zap logger [docs](https://pkg.go.dev/go.uber.org/zap#Logger.Sync),

> Applications should take care to call Sync before exiting

We have that here: https://github.com/caraml-dev/turing/blob/main/engines/router/missionctl/server/application.go#L38

I think we can remove the other instances of `.Sync()` where we are logging errors (unless there are test cases that are validating that which will require an immediate flush, which doesn't seem to be the case).